chocolatey/choco
# (#508) Fix issues when resolving dependencies and connecting to sources

## Description Of Changes

This pull request fixes an issue where failure on a source when resolving dependencies would stop the entire execution of an installation or upgrade.
Additionally, as it was required when fixing this issue, and to prevent us from blocking ourself in future related issues some refactoring was made to create a new class that holds and lazily loads any resources that we make use of when connecting to different source endpoints and replace all uses of the previously generated tuples to instead look up any cached resources (or create new resource group when the source hasn't been used yet).

## Motivation and Context

To make sure installations succeed, even if there is a problem with a single unrelated source.

## Testing

1. Enable the Chocolatey source.
2. Add a new source for the licensed repository (to make sure queries still succeed here).
3. Add another invalid source by running `choco source add -n invalid -s https://invalid.chocolatey.org/api/v2/`
4. Run `choco search chocolatey`
5. Validate there is a warning about not being able to load resource index, and still shows up the result of the search.
6. Run `choco info chocolateygui`
7. Do same validation as item 5
8. Run `choco install vlc.portable --version 3.0.16`
9. Validate same warning was outputted as item 5, but installed successfully together with the dependency
10. Run `choco upgrade vlc.portable`
11. Validate same warning was outputted as item 5, but upgraded successfully.
12. Uninstall packages, and rerun step 4-11 using `--disable-repository-optimizations`
13. Run test kitchen tests using the tests in this PR.

### Operating Systems Testing

- Windows 10/11

## Change Types Made

* [x] Breaking change (fix or feature that could cause existing functionality to change) (public methods has changed, should not be a problem as these is believed to have been introduced in 2.0.0).

## Change Checklist

* [x] All new and existing tests passed? (E2E tests not run)

## Related Issue

- Ref #508
- https://app.clickup.com/t/20540031/PROJ-531
